src/tiledimage.js

```diff
@@ -1762,80 +1770,94 @@ $.extend($.TiledImage.prototype, $.EventSource.prototype, /** @lends OpenSeadrag
+            // freedom to decide on the cache creation (note, tiles now MUST have cache, e.g.
+            // it is no longer possible to store all tiles in the memory as it was with context2D prop)
+            tile.save();
         }
 
-        const fallbackCompletion = getCompletionCallback();
         /**
          * Triggered when a tile has just been loaded in memory. That means that the
          * image has been downloaded and can be modified before being drawn to the canvas.
+         * This event awaits its handlers - they can return promises, or be async functions.
          *
-         * @event tile-loaded
+         * @event tile-loaded awaiting event
          * @memberof OpenSeadragon.Viewer
          * @type {object}
          * @property {Image|*} image - The image (data) of the tile. Deprecated.
-         * @property {*} data image data, the data sent to ImageJob.prototype.finish(), by default an Image object
+         * @property {*} data image data, the data sent to ImageJob.prototype.finish(),
+         *   by default an Image object. Deprecated
          * @property {String} dataType type of the data
          * @property {OpenSeadragon.TiledImage} tiledImage - The tiled image of the loaded tile.
          * @property {OpenSeadragon.Tile} tile - The tile which has been loaded.
          * @property {XMLHttpRequest} tileRequest - The AJAX request that loaded this tile (if applicable).
-         * @property {function} getCompletionCallback - A function giving a callback to call
-         * when the asynchronous processing of the image is done. The image will be
-         * marked as entirely loaded when the callback has been called once for each
-         * call to getCompletionCallback.
+         * @property {OpenSeadragon.Promise} - Promise resolved when the tile gets fully loaded.
+         * @property {function} getCompletionCallback - deprecated
          */
-        this.viewer.raiseEvent("tile-loaded", {
+        const promise = this.viewer.raiseEventAwaiting("tile-loaded", {
             tile: tile,
             tiledImage: this,
             tileRequest: tileRequest,
+            promise: finishPromise,
             get image() {
-                $.console.error("[tile-loaded] event 'image' has been deprecated. Use 'data' property instead.");
+                $.console.error("[tile-loaded] event 'image' has been deprecated. Use 'tile.getData()' instead.");
                 return data;
             },
-            data: data,
-            dataType: dataType,
-            getCompletionCallback: getCompletionCallback
+            get data() {
+                $.console.error("[tile-loaded] event 'data' has been deprecated. Use 'tile.getData()' instead.");
+                return data;
+            },
+            getCompletionCallback: function () {
+                $.console.error("[tile-loaded] getCompletionCallback is not supported: it is compulsory to handle the event with async functions if applicable.");
```

Comment:
This isn't a deprecation; this is complete removal. Would it be possible to deprecate it (have it still work but produce an error) for a few releases before we remove it, or is that not going to work because of how things are changing?

I'm concerned about the new version of OSD breaking a bunch of plugins. Of course some of the popular plugins that would be affected (like https://github.com/usnistgov/OpenSeadragonFiltering) aren't in active development, so if we are removing this API eventually, we may need to fork them and fix them ourselves. Regardless, no doubt people's apps use this API and it would be nice to give them a grace period if possible.

Reply:
I think this was impossible (in a nice way) since handling of this was fundamentally different from how it works now. Since its behavior was also erroneous in certain use cases, I decided for removal. I will try to look into it once again, maybe introduce switch flag that would enable the new feature and later turn that flag on by default.

As for plugins, I wanted to fork the filtering plugin and fix it to move from all the old 'dirty' ways (e.g. context2D prop). Doing so would also allow me to analyze the new system performance.

Reply:
Cool. It would be nice to have such a flag for a release or two if possible, for people's own projects. For plugins I agree it would be great to fork the filtering plugin and bring it up to date if you're up for it!

Reply:
Ok, I remembered it wrongly, the support for it can be left there, I just did not like it since the async way is much cleaner code, and not buggy. But I added it (locally for now). Also:

https://github.com/Aiosa/OpenSeadragonFiltering/blob/f1e350c08e8bf050fdc84b89e9b04ad041c861b9/openseadragon-filtering.js#L106

Even in this popular plugin it's used incorrectly. If you get a callback, you should call it, otherwise the pipeline will never finish. I mean right, they probably reset the whole tiled image and want to prevent processing things that were loaded before the reset, but I am not sure what happens with the unfinished call... 'leaks?' hanging pieces of memory, closures, or will it get freed, dereferenced? Since I am not sure I guess nobody really is.

It also seems the zombie cache will be much of use here:
https://github.com/Aiosa/OpenSeadragonFiltering/blob/f1e350c08e8bf050fdc84b89e9b04ad041c861b9/openseadragon-filtering.js#L195

Reply:
Great! And yeah, I agree it needs to be moved away from. I just want us to do it more gradually so people have a chance to update with warning :)